Evaluate on subset of entities

mberr opened the pull request:
_Transferred from the development repo._

This PR adds entity/relation filtering to the evaluation methods, i.e. given a set of entities of interest `EI`, and a set of relations of interest `RI`, it restricts the evaluation to triples `(h,r,t) in EI x RI x EI`, and considers in the ranking only the entities from `EI`.

From the implementation it evaluates only on a subset of the triples, but still computes scores for all entities to avoid irregular access patterns.

Pro:
* No irregular access pattern

Con:
* For small subset of entities, the overhead might be significant

TODOs:

* [x] add usage example to module docstring (as requested by @cthoyt )
* [x] Investigate how to propagate restrictions to the size-probing methods.
* [x] Filter triples from TriplesFactory.
* [x] <del>Add triple filtering to early stopper</del> This is automatically done, if we provide a restricted `validation_triples_factory`
* [x] Add triple filtering to pipeline


Partially solves missing documentation of RankBasedEvaluator, as requested by #19

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -53,29 +42,30 @@ class BaseModule(nn.Module):
-    def __init__(self, config: Union[Dict, BaseConfig]) -> None:
+    def __init__(self, margin_loss, num_entities, num_relations, embedding_dim,
+                 random_seed: Optional[int] = None,
+                 preferred_device: Optional[str] = 'cpu') -> None:
```
Should be set to none and default behavior should be defined inside the __init__ function

lvermue replied:
Wouldn't that just obfuscate the actual default?

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
+            np.random.seed(seed=self.random_seed)
+            torch.manual_seed(seed=self.random_seed)
+
+        if self.model_name == 'CONV_E_NAME':
```
isn't the optimizer also specified by the user? Also this sort of if statement should be taken care of by polymorphism (make the ConvE class override this part with different behavior rather than hard coding it into the base class)

lvermue replied:
It would make sense to allow the user to chose the optimizer, but in the current package, it is hard coded.

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
+        else:
+            optimizer = optim.SGD(self.parameters(), lr=learning_rate)
+
+        log.info('****Run Model On %s****' % str(self.device).upper())
```
Use f-strings

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
+
+            for i, pos_batch in enumerate(pos_batches):
+                current_batch_size = len(pos_batch)
+                batch_subjs = pos_batch[:, 0:1]
```
there's a helper function for these three lines

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
     return h, r, t
+
+def _split_list_in_batches(input_list, batch_size):
+    return [input_list[i:i + batch_size] for i in range(0, len(input_list), batch_size)]
```
- no redundant 0 in range
- needs type annotations
- needs trailing new line

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
+        self.to(self.device)
+
+        if self.random_seed is not None:
+            np.random.seed(seed=self.random_seed)
```
do we use the python random module anywhere?

lvermue replied:
Yes, in the hpo module.
I'll add it the python random module as well.

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
+            batch_size,
+            tqdm_kwargs: Optional[Mapping[str, Any]] = None,
+    ) -> List[float]:
+        """"""
```
Needs docstring

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -91,10 +81,107 @@ def __init_subclass__(cls, **kwargs):  # noqa: D105
+        self.to(self.device)
+        torch.cuda.empty_cache()
+
+    def fit(
```
needs type annotations

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -53,29 +42,30 @@ class BaseModule(nn.Module):
+                 preferred_device: Optional[str] = 'cpu') -> None:
         super().__init__()
 
-        if not isinstance(config, BaseConfig):
```
I never really liked the BaseConfig solution, but how do you suggest we load/validate configurations if not using these dataclasses?

lvermue replied:
These are now checked by the attributes of the __init__ constructor for each model.

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -4,46 +4,35 @@
-            embedding_dimension=config[EMBEDDING_DIM],
-        )
 
+def get_device(device):
```
This discarded the logic for whether the user would like to use GPU or not...

lvermue replied:
It didn't as far as I'm concerned.
The preferred device is still given to the kge module and the function checks if cuda is available in case the user would like to use the GPU.

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -4,46 +4,35 @@
     'BaseModule',
-    'BaseConfig',
-    'slice_triples',
+    'slice_triples'
```
always use trailing commas to reduce diffs

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -85,24 +76,152 @@ def __init__(self, config: Union[Dict, BaseConfig]) -> None:
+                self.device = torch.device('cuda')
+            else:
+                self.device = torch.device('cpu')
+                print('No cuda devices were available. The model runs on CPU')
```
No prints, use logging

cthoyt commented on src/pykeen/kge_models/base.py:
```diff
@@ -85,24 +76,152 @@ def __init__(self, config: Union[Dict, BaseConfig]) -> None:
+            tqdm_kwargs: Optional[Mapping[str, Any]] = None,
+    ) -> List[float]:
+        """
+        Trains the kge model with the given parameters
```
Use correct doc8 style (imperative mood, on first line)

cthoyt commented on src/pykeen/kge_models/conv_e.py:
```diff
@@ -33,74 +36,69 @@ class ConvE(nn.Module):
+                 num_entities: int,
+                 num_relations: int,
+                 embedding_dim: int,
+                 ConvE_input_channels, ConvE_output_channels, ConvE_height, ConvE_width, ConvE_kernel_height,
```
Switch to pep-8 compliant variable names (see flake8 - the build will give all of these hints for code style)

mberr committed:
Add entity restriction option to RankBasedEvaluator

mberr committed:
Add docstring

mberr committed:
Adhere to docstring convention

mberr committed:
Add relation/entity restriction to evaluate(...)

mberr committed:
Add relation/entity restriction to Evaluator.evaluate(...)

mberr committed:
Remove entity restriction from RankBasedEvaluator

This functionality is now provided in the general Evaluator class

mberr committed:
Fix "f-string is missing placeholder"

mberr committed:
Try to fix D401

mberr committed:
Refactor evaluate to reduce complexity

mberr committed:
Fix type annotation

mberr committed:
Fix undefined variable

mberr committed:
Fix variable names and comments

The generic evaluate_ method was still using the
tail-centered naming convention

mberr committed:
Improve memory efficiency

mberr committed:
Add filter triples method to TriplesFactory

mberr committed:
Move filter code outside of evaluate

mberr committed:
Pas restrict_entities_to to size probing

mberr committed:
Update type annotation and docstring of batch_and_slice

mberr committed:
Remove dead code

mberr committed:
Add NotImplementedError

mberr committed:
Support inverse relations

cthoyt committed:
Merge branch 'master' into evaluate-on-subset-of-entities

mberr committed:
Remove outdated TODO

mberr committed:
Merge remote-tracking branch 'origin/master' into evaluate-on-subset-of-entities

# Conflicts:
#	src/pykeen/evaluation/evaluator.py

mberr committed:
Add flag for time-consuming input checks

mberr committed:
Add flag for time-consuming input checks

mberr committed:
Add base-case for empty triples

mberr committed:
Fix new_with_restrictin when no relation restriction is provided

mberr committed:
Add unittest for TriplesFactory.new_with_restriction

mberr committed:
Merge remote-tracking branch 'pykeen-public/master' into evaluate-on-subset-of-entities

# Conflicts:
#	src/pykeen/triples/triples_factory.py

mberr committed:
Fix typo

mberr committed:
Fix over-indention

mberr committed:
Fix merge artifact

mberr committed:
Revert change

cthoyt commented on src/pykeen/evaluation/evaluator.py:
```diff
@@ -449,12 +469,31 @@ def evaluate(
 
+    # verify that the triples have been filtered
+    if restrict_entities_to is not None and do_time_consuming_checks:
+        present_entity_ids = set(mapped_triples[:, 0].unique().tolist()).union(mapped_triples[:, 2].unique().tolist())
```
do we have a utility somewhere that gets the entity list from a set of triples? We've rewritten this line so many times

mberr replied:
6e294b3

mberr committed:
Extract utility method

mberr committed:
Revert white-space change

mberr committed:
Add evaluation restriction to pipeline

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -580,11 +582,19 @@ def pipeline(  # noqa: C901
         The keyword arguments passed to the dataset upon instantiation
     :param training_triples_factory:
-        A triples factory with training instances if a a dataset was not specified
+        A triples factory with training instances if a dataset was not specified.
```
not a sentence -> no period at the end

mberr committed:
Fix error for datasets without validation_triples_factory

mberr committed:
Remove period after incomplete sentences

mberr committed:
Fix typo

mberr committed:
Fix inconsistent indention

mberr committed:
Merge remote-tracking branch 'origin/master' into evaluate-on-subset-of-entities

mberr committed:
Add first draft of entity/relation restriction doc

mberr committed:
Merge remote-tracking branch 'origin/master' into evaluate-on-subset-of-entities

# Conflicts:
#	src/pykeen/triples/triples_factory.py
#	tests/test_triples_factory.py

cthoyt committed:
Pass doc8

cthoyt commented on docs/source/tutorial/understanding_evaluation.rst:
```diff
@@ -70,10 +70,27 @@ incapability to solve of one the tasks.
+
+1. We filter the evaluation triples :math:`\mathcal{T}_{eval}` to contain only those triples which are of interest, i.e.
+   :math:`\mathcal{T}_{eval}' = \{(h, r, t) \in \mathcal{T}_{eval} \mid h, t \in \mathcal{E}_I, r \in \mathcal{R}_I\}`
+2. We restrict the candidate entities :math:`e'` for a triple :math:`(h, r, t)` to :math:`e' \in \mathcal{E}_{eval}`.
```
This is a bit confusing. I know what it's supposed to say, but an unfamiliar reader might not get the idea that the candidate entity e' is either the t' in (h,r,t') triples for tail prediction or the h' in (h',r,t) triples for head prediction.

mberr replied:
a3012c4

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -531,6 +531,8 @@ def pipeline(  # noqa: C901
     training_triples_factory: Optional[TriplesFactory] = None,
     testing_triples_factory: Optional[TriplesFactory] = None,
     validation_triples_factory: Optional[TriplesFactory] = None,
+    restrict_evaluation_to_entities: Optional[Collection[str]] = None,
```
maybe `evaluation_entity_whitelist` might be more descriptive. Same for the relations on the next line

mberr replied:
fc26d54

cthoyt commented on src/pykeen/pipeline.py:
```diff
@@ -663,6 +674,17 @@ def pipeline(  # noqa: C901
         validation_triples_factory=validation_triples_factory,
     )
 
+    # evaluation restriction to a subset of entities/relations
```
if  the restriction list is none, it should just pass through (just in case there's an issue with `TriplesFactory.new_with_restriction()`)

cthoyt replied:
Something like this might work:

```python
    testing_triples_factory, validation_triples_factory = [
        None if factory is None else
        factory.new_with_restriction(
            entities=restrict_evaluation_to_entities,
            relations=restrict_evaluation_to_relations,
        ) if restrict_evaluation_to_entities or restrict_evaluation_to_relations else
        factory
        for factory in (testing_triples_factory, validation_triples_factory)
    ]
```

mberr replied:
5e08e34

cthoyt commented on src/pykeen/triples/triples_factory.py:
```diff
@@ -125,6 +130,10 @@ def _map_triples_elements_to_ids(
     relation_to_id: RelationMapping,
 ) -> MappedTriples:
     """Map entities and relations to pre-defined ids."""
+    if triples.size < 1:
```
can't this just be a zero check? Are negative sizes possible? 😛

mberr replied:
9d34aa5 (although functionally the same) :wink:

cthoyt committed:
No slashes

[skip ci]

cthoyt committed:
Update test_triples_factory.py

[skip ci]

mberr committed:
Update understanding_evaluation.rst

mberr committed:
Rename parameter

mberr committed:
Change restricting factories to entities/relations of interest

mberr committed:
Change check for empty triples

cthoyt committed:
Update text

cthoyt committed:
Add example usage

[skip ci]

cthoyt reviewed (approved):
I fixed the linting errors with the documentation and added an example. Assuming this passes Travis, it's ready to merge! Thanks a ton @mberr

cthoyt committed:
Merge branch 'master' into evaluate-on-subset-of-entities

cthoyt committed:
Pass flake8
